File: internal/pkg/act/net_decoder.go
/*
This file is part of FFLiveParse.

FFLiveParse is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

FFLiveParse is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with FFLiveParse.  If not, see <https://www.gnu.org/licenses/>.
*/

package act

import (
	"bytes"
	"compress/gzip"
	"encoding/binary"
	"errors"
	"net"
	"time"

	"../app"
)

func readUint16(data []byte, pos *int) uint16 {
	if len(data)-*pos < 2 {
		return 0
	}
	dataString := data[*pos : *pos+2]
	*pos += 2
	return binary.BigEndian.Uint16(dataString)
}

func readUint32(data []byte, pos *int) uint32 {
	if len(data)-*pos < 4 {
		return 0
	}
	dataString := data[*pos : *pos+4]
	*pos += 4
	return binary.BigEndian.Uint32(dataString)
}

func readInt32(data []byte, pos *int) int32 {
	return int32(readUint32(data, pos))
}

func readByte(data []byte, pos *int) byte {
	if len(data)-*pos < 1 {
		return 0
	}
	output := data[*pos]
	*pos++
	return output
}

func readString(data []byte, pos *int) string {
	length := int(readUint16(data, pos))
	if length == 0 || len(data)-*pos < length {
		return ""
	}
	output := string(data[*pos : *pos+length])
	*pos += length
	return output
}

func readTime(data []byte, pos *int) time.Time {
	timeString := readString(data, pos)
	time, _ := time.Parse(time.RFC3339, timeString)
	return time
}

// DecodeSessionBytes - Create Session struct from incomming data packet
func DecodeSessionBytes(data []byte, addr *net.UDPAddr) (Session, int, error) {
	if data[0] != DataTypeSession {
		return Session{}, 0, errors.New("invalid data type for Session")
	}
	pos := 1
	// check version number
	versionNumber := readInt32(data, &pos)
	if versionNumber < app.ActPluginMinVersionNumber || versionNumber > app.ActPluginMaxVersionNumber {
		return Session{}, 0, errors.New("version number mismatch")
	}
	return Session{
		UploadKey: readString(data, &pos),
		IP:        addr.IP,
		Port:      addr.Port,
	}, pos, nil
}

// DecodeEncounterBytes - Create Encounter struct from incomming data packet
func DecodeEncounterBytes(data []byte) (Encounter, int, error) {
	if data[0] != DataTypeEncounter {
		return Encounter{}, 0, errors.New("invalid data type for Encounter")
	}
	pos := 1
	return Encounter{
		ActID:        readUint32(data, &pos),
		StartTime:    readTime(data, &pos),
		EndTime:      readTime(data, &pos),
		Zone:         readString(data, &pos),
		Damage:       readInt32(data, &pos),
		Active:       readByte(data, &pos) != 0,
		SuccessLevel: readByte(data, &pos),
	}, pos, nil
}

// DecodeCombatantBytes - Create Combatant struct from incomming data packet
func DecodeCombatantBytes(data []byte) (Combatant, int, error) {
	if data[0] != DataTypeCombatant {
		return Combatant{}, 0, errors.New("invalid data type for Combatant")
	}
	pos := 1
	actEncounterID := readUint32(data, &pos)
	p := Player{
		ID:   readInt32(data, &pos),
		Name: readString(data, &pos),
	}
	p.ActName = p.Name
	c := Combatant{
		Player:         p,
		ActEncounterID: actEncounterID,
		Job:            readString(data, &pos),
		Damage:         readInt32(data, &pos),
		DamageTaken:    readInt32(data, &pos),
		DamageHealed:   readInt32(data, &pos),
		Deaths:         readInt32(data, &pos),
		Hits:           readInt32(data, &pos),
		Heals:          readInt32(data, &pos),
		Kills:          readInt32(data, &pos),
		Time:           time.Now(),
	}
	return c, pos, nil
}

// DecodeLogLineBytes - Create LogLine struct from incomming data packet
func DecodeLogLineBytes(data []byte) (LogLine, int, error) {
	if data[0] != DataTypeLogLine {
		return LogLine{}, 0, errors.New("invalid data type for LogLine")
	}
	pos := 1
	encounterID := readUint32(data, &pos)
	time := readTime(data, &pos)
	logLine := readString(data, &pos)
	return LogLine{
		ActEncounterID: encounterID,
		Time:           time,
		LogLine:        logLine,
	}, pos, nil
}

// DecompressBytes - Decompress byte array for recieving
func DecompressBytes(data []byte) ([]byte, error) {
	r := bytes.NewReader(data)
	gz, err := gzip.NewReader(r)
	defer gz.Close()
	var output bytes.Buffer
	_, err = output.ReadFrom(gz)
	if err != nil {
		return nil, err
	}
	return output.Bytes(), nil
}

// DecodeLogLineBytesFile - Create LogLine struct from data stored in log file
func DecodeLogLineBytesFile(data []byte) ([]LogLine, int, error) {
	// should be compressed
	logBytes, err := DecompressBytes(data)
	if err != nil {
		return nil, 0, err
	}
	// itterate log bytes and convert to log line
	pos := 0
	logLines := make([]LogLine, 0)
	for pos < len(logBytes) {
		// check 'type' byte
		if logBytes[pos] != DataTypeLogLine {
			return nil, 0, errors.New("invalid data type for LogLine")
		}
		// read data
		pos = pos + 1
		encounterUID := readString(logBytes, &pos)
		time := readTime(logBytes, &pos)
		logLineString := readString(logBytes, &pos)
		// append to log lines array
		logLines = append(
			logLines,
			LogLine{
				EncounterUID: encounterUID,
				Time:         time,
				LogLine:      logLineString,
			},
		)
	}
	return logLines, pos, nil
}

File: web/static/js/worker/net_decoder.js
/*
This file is part of FFLiveParse.

FFLiveParse is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

FFLiveParse is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with FFLiveParse.  If not, see <https://www.gnu.org/licenses/>.
*/

var DATA_TYPE_ENCOUNTER = 2;
var DATA_TYPE_COMBATANT = 3;
var DATA_TYPE_LOG_LINE = 5;
var DATA_TYPE_FLAG = 99;

var SIZE_BYTE = 1;
var SIZE_INT16 = 2;
var SIZE_INT32 = 4;

var totalBytesRecieved = 0;
var hasEncounter = false;
var reportStatus = false;
var partialBuffer = {};

function readInt32(data, pos)
{
    return new DataView(data.buffer, pos, SIZE_INT32).getInt32(0)
}

function readUint32(data, pos)
{
    return new DataView(data.buffer, pos, SIZE_INT32).getUint32(0)
}

function readUint16(data, pos)
{
    return new DataView(data.buffer, pos, SIZE_INT16).getUint16(0)
}

function readByte(data, pos)
{
    return data[pos];
}

function readString(data, pos)
{
    var strLen = readUint16(data, pos);
    return new TextDecoder("utf-8").decode(data.slice(pos + 2, pos + 2 + strLen));
}

// @see https://stackoverflow.com/questions/40031688/javascript-arraybuffer-to-hex
function buf2hex(buffer)
{
    return Array.prototype.map.call(new Uint8Array(buffer), x => ('00' + x.toString(16)).slice(-2)).join('');
}

// @see https://stackoverflow.com/questions/33702838/how-to-append-bytes-multi-bytes-and-buffer-to-arraybuffer-in-javascript
function concatTypedArrays(a, b) {
    var c = new (a.constructor)(a.length + b.length);
    c.set(a, 0);
    c.set(b, a.length);
    return c;
}

function decodeEncounterBytes(data)
{
    if (data[0] != DATA_TYPE_ENCOUNTER) {
        return null;
    }
    var pos = 1;
    var output = {
        "Type" : DATA_TYPE_ENCOUNTER
    };
    output["UID"]           = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["StartTime"]     = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["EndTime"]       = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["Zone"]          = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["Damage"]        = readUint32(data, pos); pos += SIZE_INT32;
    output["Active"]        = readByte(data, pos) != 0; pos += SIZE_BYTE;
    output["EndWait"]       = readByte(data, pos) != 0; pos += SIZE_BYTE;
    output["SuccessLevel"]  = readByte(data, pos); pos += SIZE_BYTE;
    
    output["StartTime"]     = new Date(output["StartTime"]);
    output["EndTime"]       = new Date(output["EndTime"]);

    if (!encounterUid || output["UID"] == encounterUid) {
        if (output.Zone) {
            hasEncounter = true;
        }
        postMessage({
            "type"      : "act:encounter",
            "data"      : output
        });
    }
    return pos;
}

function decodeCombatantBytes(data)
{
    if (data[0] != DATA_TYPE_COMBATANT) {
        return 0;
    }
    var pos = 1;
    var output = {
        "Type" : DATA_TYPE_COMBATANT
    };
    output["EncounterUID"]  = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["ID"]            = readInt32(data, pos); pos += SIZE_INT32;
    output["Name"]          = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["World"]         = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["Job"]           = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["Damage"]        = readInt32(data, pos); pos += SIZE_INT32;
    output["DamageTaken"]   = readInt32(data, pos); pos += SIZE_INT32;
    output["DamageHealed"]  = readInt32(data, pos); pos += SIZE_INT32;
    output["Deaths"]        = readInt32(data, pos); pos += SIZE_INT32;
    output["Hits"]          = readInt32(data, pos); pos += SIZE_INT32;
    output["Heals"]         = readInt32(data, pos); pos += SIZE_INT32;
    output["Kills" ]        = readInt32(data, pos); pos += SIZE_INT32;
    output["Time"]          = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["Time"]          = new Date(output["Time"]);
    if (!encounterUid || output["EncounterUID"] == encounterUid) {
        postMessage({
            "type"      : "act:combatant",
            "data"      : output
        });
    }
    return pos;
}

function decodeLogLineBytes(data)
{
    if (data[0] != DATA_TYPE_LOG_LINE) {
        return 0;
    }
    var pos = 1;
    var output = {
        "Type" : DATA_TYPE_LOG_LINE
    };
    output["EncounterUID"]  = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["Time"]          = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["LogLine"]       = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;

    output["Time"]          = new Date(output["Time"]);
    if (!encounterUid || output["EncounterUID"] == encounterUid) {
        postMessage({
            "type"      : "act:logLine",
            "data"      : output
        });
    }
    return pos;
}

function decodeFlagBytes(data)
{
    if (data[0] != DATA_TYPE_FLAG) {
        return 0;
    }
    var pos = 1;
    var output = {
        "Type" : DATA_TYPE_FLAG
    };
    output["Name"]          = readString(data, pos); pos += readUint16(data, pos) + SIZE_INT16;
    output["Value"]         = readByte(data, pos) != 0; pos += SIZE_BYTE;
    postMessage({
        "type"      : "flag",
        "data"      : output
    });
    return pos;
}

function parseNextMessage(buffer, pos, count)
{
    var data = new Uint8Array(buffer.slice(pos));
    var length = 0;
    switch (data[0])
    {
        case DATA_TYPE_ENCOUNTER:
        {
            length = decodeEncounterBytes(data);
            break;
        }
        case DATA_TYPE_COMBATANT:
        {
            length = decodeCombatantBytes(data);
            break;
        }
        case DATA_TYPE_LOG_LINE:
        {
            length = decodeLogLineBytes(data);
            break;
        }
        case DATA_TYPE_FLAG:
        {
            length = decodeFlagBytes(data);
            break;
        }
    }
    delete data;
    if (length <= 0) {
        postMessage({
            "type"      : "status_ready"
        });
        return;
    }
    pos += length;
    // every ~200 send status update and use setTimeout to prevent recursion error
    if (count > 200) {
        if (reportStatus) {
            var loadingProgress = ((pos / buffer.byteLength) * 100).toFixed(1);
            postMessage({
                "type"      : "status_in_progress",
                "message"   : "Loading (" + loadingProgress + "%)",
                "value"     : loadingProgress
            });
        }
        setTimeout(function() { parseNextMessage(buffer, pos, 0); }, 1);
        return;
    }
    parseNextMessage(buffer, pos, count + 1);
}

function parseMessage(data)
{
    totalBytesRecieved += data.length;
    reportStatus = false;
    if (data.length > 102400) {
        reportStatus = true;
    }
    // decompress
    data = pako.inflate(data)
    if (data.length > 0) {
        parseNextMessage(data.buffer, 0, 0);
    }
}
